doc/cprover-manual/goto-instrument.md

```diff
@@ -0,0 +1,127 @@
+[CPROVER Manual TOC](../../)
+
+## Goto Instrument
+
+This is a tool that implements a variety of instrumentation passes modifying a
+pre-existing GOTO binary, or rather producing the modified binary as output.
+
+```sh
+$ goto-instrument in.gb out.gb <instrumentation-options>
+```
+
+### Diagnosis Options
+
+### Safety Checks
+
+### Semantic Transformations
+
+#### Generate Function Body
+
+This transformation inserts implementation of functions without definition, i.e.
+a body. The user is allowed to choose the behavior to be implemented by those
+functions from a predefined list. The command-line option to select the behavior
+is `--generate-function-body-options <option>`.
+
+- `assert-false`: The body consists of a single command: `assert(0);`
+- `assume-false`: The body consists of a single command: `assume(0);`
+- `nondet-return`: The body only returns a non-deterministic value of it's
+  return type.
+- `assert-false-assume-false`: Two commands as above.
+- `havoc[,params:<p-regex>][,globals:<g-regex>]`: Assign non-deterministic
+  values to the targets of pointer-to-non-constant parameters matching the
+  `p-regex`, and non-constant globals matching the `g-regex` and then (in case
+  of non-void function) returning as with `nondet-return`.
+
+Let's demonstrate using the havoc on a simple example:
+
+```C
+// main.c
+int global;
+const int c_global;
+
+int function(int *param, const int *c_param);
+```
+
+Often we want to avoid overwriting internal symbols, i.e. those with `__`
+prefix, which is done using the pattern `(?!__)`.
+
+```sh
+$ goto-cc main.c -o main.gb
+$ goto-instrument main.gb main-out.gb \
+  --generate-function-body-options 'havoc,params:(?!__).*,globals:(?!__).*' \
+  --generate-funtion-body function
+```
+
+Leads to a binary equivalent to the following C code:
+
+```C
+// main-mod.c
+int function(int *param, const int *c_param) {
+  *param = nondet_int();
+  global = nondet_int();
+  return nondet_int();
+}
+```
+
+Which parameters should be modified can be specified either by a regular
+expression (as above) or by a semicolon-separated list of their numbers. For
+example `havoc,params:0;3;4` will assign non-deterministic values to the 1st,
+4th, and 5th parameter.
```

Comment:
Is that 1st, 4th and 5th parameters regardless of type, or 1st, 4th and 5th *pointer* parameters? e.g. is it type aware or not?

Reply:
So the implementation is type-safe in that it silently ignores the user-selected numbers if the respective parameter is not a pointer. Neither does it produce any error even if the user number parameter is higher than the number of parameters a given function takes. Do you think it should explicitly fail, or should we just extend the documentation?

Reply:
From an end-user point of view "silently ignore and not do what I meant" is rather annoying - so I actually think both should be done - i.e. documentation should be extended a bit, and then the tool should give an explicit error if the user has specified something that "doesn't make sense".

Reply:
Done.